README.md
# [Okaeri](https://gitlab.com/GCSBOSS/okaeri)

Every standard functionallity regarding account management and authentication in one library

## Reporting Bugs
If you have found any problems with this module, please:

1. [Open an issue](https://gitlab.com/GCSBOSS/okaeri/issues/new).
2. Describe what happened and how.
3. Also in the issue text, reference the label `~bug`.

We will make sure to take a look when time allows us.

## Proposing Features
If you wish to get that awesome feature or have some advice for us, please:
1. [Open an issue](https://gitlab.com/GCSBOSS/okaeri/issues/new).
2. Describe your ideas.
3. Also in the issue text, reference the label `~proposal`.

## Contributing
If you have spotted any enhancements to be made and is willing to get your hands
dirty about it, fork us and
[submit your merge request](https://gitlab.com/GCSBOSS/okaeri/merge_requests/new)
so we can collaborate effectively.


CHANGELOG.md
# Okaeri Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [v0.1.4] - 2021-03-12

### Added
- auth date field to account to be updated whenever its credentials are checked

## [v0.1.3] - 2021-03-12

### Fixed
- accounts query always including id ordering
- groups query always including id ordering

## [v0.1.2] - 2021-03-04

### Added
- function to check whether an account belongs to any given groups

### Fixed
- `accounts` field from group not returned correctly on `readGroup()`

## [v0.1.1] - 2021-03-03

### Added
- account group functionality
- account iteration feature

## [v0.1.0] - 2020-12-10

### Changed
- the entire project and API to a whole new direction

## [v0.0.6] - 2020-11-24

### Fixed
- add missing TOML dependency

## [v0.0.5] - 2020-11-24

### Fixed
- reduced footprint due to upgrade of nodecaf version

## [v0.0.4] - 2020-04-17

### Fixed
- deployment process along with nodecaf version

## [v0.0.3] - 2020-04-02

### Added
- endpoints to fetch account data based on header and on parameter

### Changed
- create and auth endpoints to respond with account id

## [v0.0.2] - 2020-02-13

### Added
- mongodb unique index to account name

### Fixed
- error of missing config file when running from docker
- docker exposed port
- warning from mongo client settings

### Changed
- logging system to new nodecaf (12 factor) stdout logging

## [v0.0.1] - 2020-01-18

First officially released version

[v0.0.1]: https://gitlab.com/GCSBOSS/okaeri/-/tags/v0.0.1
[v0.0.2]: https://gitlab.com/GCSBOSS/okaeri/-/tags/v0.0.2
[v0.0.3]: https://gitlab.com/GCSBOSS/okaeri/-/tags/v0.0.3
[v0.0.4]: https://gitlab.com/GCSBOSS/okaeri/-/tags/v0.0.4
[v0.0.5]: https://gitlab.com/GCSBOSS/okaeri/-/tags/v0.0.5
[v0.0.6]: https://gitlab.com/GCSBOSS/okaeri/-/tags/v0.0.6
[v0.1.0]: https://gitlab.com/GCSBOSS/okaeri/-/tags/v0.1.0
[v0.1.1]: https://gitlab.com/GCSBOSS/okaeri/-/tags/v0.1.1
[v0.1.2]: https://gitlab.com/GCSBOSS/okaeri/-/tags/v0.1.2
[v0.1.3]: https://gitlab.com/GCSBOSS/okaeri/-/tags/v0.1.3
[v0.1.4]: https://gitlab.com/GCSBOSS/okaeri/-/tags/v0.1.4


test/spec.js
const assert = require('assert');
const Okaeri = require('../lib/main');

process.env.NODE_ENV = 'testing';

const mongoConf = {
    url: 'mongodb://' + (process.env.MONGO_HOST || 'localhost:27017'),
    name: 'Okaeri'
};

let okaeri, mongo = null;

beforeEach(async function(){

    okaeri = new Okaeri({ mongo: mongoConf, identityHeader: 'id' });
    await okaeri.start();
    mongo = okaeri.mongo;
});

afterEach(async function(){
    await mongo.account.deleteOne();
    await mongo.group.deleteOne();
    await okaeri.stop();
});

describe('Okaeri', function(){

    describe('Creating an account', function(){

        it('Should fail when input is invalid', async function(){
            let res = await okaeri.createAccount({ password: 'blah' });
            assert(!res.ok);
        });

        it('Should create account in database', async function(){
            assert(!await mongo.account.findOne({ name: 'test-a' }));
            let res = await okaeri.createAccount({ name: 'test-a', password: 'foobarbaz' });
            assert(res.ok);
            assert(await mongo.account.findOne({ name: 'test-a' }));
        });

        it('Should fail when account already exists', async function(){
            await okaeri.createAccount({ name: 'test-a', password: 'foobarbaz' });
            let res = await okaeri.createAccount({ name: 'test-a', password: 'foobarbaz' });
            assert(!res.ok);
            assert.strictEqual(res.type, 'conflict');
        });

    });

    describe('Verifying authentication', function(){

        it('Should fail when input is invalid', async function(){
            let res = await okaeri.checkCredentials(null, 'blah');
            assert(!res.ok);
        });

        it('Should fail when account doesn\'t exist', async function(){
            let res = await okaeri.checkCredentials('test-a', 'blah');
            assert(!res.ok);
            assert.strictEqual(res.type, 'wrong');
        });

        it('Should fail when password is wrong', async function(){
            await okaeri.createAccount({ name: 'test-a', password: 'foobarbaz' });
            let res = await okaeri.checkCredentials('test-a', 'wrong-pass');
            assert(!res.ok);
            assert.strictEqual(res.type, 'wrong');
        });

        it('Should respond OK when name and password are correct', async function(){
            await okaeri.createAccount({ name: 'test-a', password: 'foobarbaz' });
            let res = await okaeri.checkCredentials('test-a', 'foobarbaz');
            assert(res.ok);
        });

    });

    describe('Fetching account info', function(){

        it('Should fail when account doesn\'t exist', async function(){
            let res = await okaeri.readAccount('abdcfe61524635124');
            assert(!res.ok);
            assert.strictEqual(res.type, 'unknown');
        });

        it('Should return account info', async function(){
            let { id } =  await okaeri.createAccount({ name: 'test-a', password: 'foobarbaz' });
            let res = await okaeri.readAccount(id);
            assert(res.ok);
            assert.strictEqual(res.acc.name, 'test-a');
        });

    });

    describe('Updating an account', function(){

        it('Should fail when account doesn\'t exist', async function(){
            let res = await okaeri.updateAccount('abdcfe61524635124');
            assert(!res.ok);
            assert.strictEqual(res.type, 'unknown');
        });

        it('Should update account in database', async function(){
            let { id } = await okaeri.createAccount({ name: 'test-a', password: 'foobarbaz' });
            let res = await okaeri.updateAccount(id, { name: 'test-z' });
            assert(res.ok);
            let r = await okaeri.readAccount(id);
            assert(r.ok);
            assert.strictEqual(r.acc.name, 'test-z');
        });

    });

    describe('Query accounts', function(){

        it('Should return all account records', async function(){
            await okaeri.createAccount({ name: 'test-a', password: 'foobarbaz' });
            await okaeri.createAccount({ name: 'test-z', password: 'foobarbaz' });
            let res = await okaeri.queryAccounts();
            assert.strictEqual(res.accs.length, 2);
        });

        it('Should filter returned accounts', async function(){
            await okaeri.createAccount({ name: 'test-a', password: 'foobarbaz' });
            await okaeri.createAccount({ name: 'test-z', password: 'foobarbaz' });
            let res = await okaeri.queryAccounts({ name: { $regex: /z/ } });
            assert.strictEqual(res.accs.length, 1);
            assert.strictEqual(res.accs[0].name, 'test-z');
        });

    });

    describe('Iterating accounts', function(){

        it('Should allow iterating through all accounts', async function(){
            await okaeri.createAccount({ name: 'test-a', password: 'foobarbaz' });
            await okaeri.createAccount({ name: 'test-z', password: 'foobarbaz' });

            let res = await okaeri.iterateAccounts();

            let a1 = await res.next();
            assert.strictEqual(a1.name, 'test-a');
            let a2 = await res.next();
            assert.strictEqual(a2.name, 'test-z');
        });

        it('Should allow iterating through accounts matched by filter', async function(){
            await okaeri.createAccount({ name: 'test-a', password: 'foobarbaz' });
            await okaeri.createAccount({ name: 'test-z', password: 'foobarbaz' });

            let res = await okaeri.iterateAccounts({ name: { $regex: /z/ } });

            let a1 = await res.next();
            assert.strictEqual(a1.name, 'test-z');
            assert(!await res.next());
        });

    });

    describe('Creating a Group', function(){

        it('Should fail when input is invalid', async function(){
            let res = await okaeri.createGroup();
            assert(!res.ok);
        });

        it('Should create a group in database', async function(){
            let res = await okaeri.createGroup({ name: 'g1', code: 'g1' });
            assert(res.ok);
            assert(await mongo.group.findOne({ name: 'g1' }));
        });

        it('Should fail when the group already exists', async function(){
            await okaeri.createGroup({ name: 'g1', code: 'g1' });
            let res = await okaeri.createGroup({ name: 'g2', code: 'g1' });
            assert(!res.ok);
            assert.strictEqual(res.type, 'conflict');
        });

    });

    describe('Updating a Group', function(){

        it('Should fail when input is invalid', async function(){
            let res = await okaeri.updateGroup('abdcfe61524635124', false);
            assert(!res.ok);
        });

        it('Should fail when group doesn\'t exist', async function(){
            let res = await okaeri.updateGroup('abdcfe61524635124');
            assert(!res.ok);
            assert.strictEqual(res.type, 'unknown');
        });

        it('Should update group in database', async function(){
            let { id } = await okaeri.createGroup({ name: 'g1', code: 'g1' });
            let res = await okaeri.updateGroup(id, { name: 'g0' });
            assert(res.ok);
            assert(await mongo.group.findOne({ name: 'g0' }));
        });

    });

    describe('Fetching Group info', function(){

        it('Should fail when group doesn\'t exist', async function(){
            let res = await okaeri.readGroup('abdcfe61524635124');
            assert(!res.ok);
            assert.strictEqual(res.type, 'unknown');
        });

        it('Should return group info', async function(){
            let { id } =  await okaeri.createGroup({ name: 'g1', code: 'g1' });
            let res = await okaeri.readGroup(id);
            assert(res.ok);
            assert.strictEqual(res.group.name, 'g1');
        });

    });

    describe('Removing group', function(){

        it('Should fail when group doesn\'t exist', async function(){
            let res = await okaeri.removeGroup('abdcfe61524635124');
            assert(!res.ok);
            assert.strictEqual(res.type, 'unknown');
        });

        it('Should remove the group from db', async function(){
            let { id } = await okaeri.createGroup({ name: 'g1', code: 'g1' });
            let res = await okaeri.removeGroup(id);
            assert(res.ok);
            assert.strictEqual(res.code, 'g1');
            let r = await okaeri.readGroup(id);
            assert(!r.ok);
        });

    });

    describe('Query groups', function(){

        it('Should return all group records', async function(){
            await okaeri.createGroup({ name: 'g1', code: 'g1' });
            await okaeri.createGroup({ name: 'g2', code: 'g2' });
            let res = await okaeri.queryGroups();
            assert.strictEqual(res.groups.length, 2);
        });

        it('Should filter returned groups', async function(){
            await okaeri.createGroup({ name: 'g1', code: 'g1' });
            await okaeri.createGroup({ name: 'g2', code: 'g2' });
            let res = await okaeri.queryGroups({ name: { $regex: /2/ } });
            assert.strictEqual(res.groups.length, 1);
            assert.strictEqual(res.groups[0].name, 'g2');
        });

    });

    describe('Manage group accounts', function(){

        it('Should fail when account doesn\'t exist', async function(){
            let r1 = await okaeri.addAccountToGroup('234235abfd5765fe', '234235abfd5765fe');
            assert(!r1.ok);

            let r2 = await okaeri.removeAccountFromGroup('234235abfd5765fe', '234235abfd5765fe');
            assert(!r2.ok);
        });

        it('Should fail when group doesn\'t exist', async function(){
            let { id } = await okaeri.createAccount({ name: 'test-a', password: 'foobarbaz' });

            let r1 = await okaeri.addAccountToGroup(id, '234235abfd5765fe');
            assert(!r1.ok);

            let r2 = await okaeri.removeAccountFromGroup(id, '234235abfd5765fe');
            assert(!r2.ok);
        });

        it('Should add account to group', async function(){
            let { id } = await okaeri.createAccount({ name: 'test-a', password: 'foobarbaz' });
            let { id: gid } = await okaeri.createGroup({ name: 'g1', code: 'g1' });
            let res = await okaeri.addAccountToGroup(id, gid);
            assert(res.ok);
            let { group } = await okaeri.readGroup(gid);
            assert.strictEqual(group.accounts[0].name, 'test-a');
        });

        it('Should remove account to group', async function(){
            let { id } = await okaeri.createAccount({ name: 'test-a', password: 'foobarbaz' });
            let { id: gid } = await okaeri.createGroup({ name: 'g1', code: 'g1' });
            await okaeri.addAccountToGroup(id, gid);
            let res = await okaeri.removeAccountFromGroup(id, gid);
            assert(res.ok);
            let { group } = await okaeri.readGroup(gid);
            assert.strictEqual(group.accounts.length, 0);
        });

    });

    describe('Check if account belongs to group', function(){

        it('Should return whether account is in any group sent', async function(){
            let { id } = await okaeri.createAccount({ name: 'test-a', password: 'foobarbaz' });
            await okaeri.createGroup({ name: 'g1', code: 'g1' });
            let { id: i2 } = await okaeri.createGroup({ name: 'g2', code: 'g2' });


            let r1 = await okaeri.isAccountInAnyGroup(id, [ 'g1', 'g2' ]);
            assert(!r1.ok);

            await okaeri.addAccountToGroup(id, i2);
            let r2 = await okaeri.isAccountInAnyGroup(id, [ 'g2' ]);
            assert(r2.ok);

            let r3 = await okaeri.isAccountInAnyGroup(id, [ 'g1', 'g2' ]);
            assert(r3.ok);
        });

    });

});


lib/account.js
const crypto = require('crypto');
const v = require('valinor');
const { post } = require('muhb');

function generateSalt(){
    return new Promise( (resolve, reject) =>
        crypto.randomBytes(32, (err, buf) =>
            /* istanbul ignore next */
            err ? reject(err) : resolve(buf.toString('hex'))));
}

function hashPassword(password, salt){
    return new Promise( (resolve, reject) =>
        crypto.pbkdf2(password, salt, 1e5, 64, 'sha512', (err, buf) =>
            /* istanbul ignore next */
            err ? reject(err) : resolve(buf.toString('hex'))));
}

module.exports = {

    async create({ mongo, res, log, body, conf }){

        let out = v.schema({
            name: v.str,
            password: v.str.between(8, 128)
        }).test(body);
        res.badRequest(!out.ok, out.errs);

        let conflict = await mongo.findOne({ name: body.name });
        res.badRequest(conflict, 'Account name \'%s\' is taken', body.name);

        let salt = await generateSalt();

        let acc = {
            salt,
            name: body.name,
            hash: await hashPassword(body.password, salt),
            creation: new Date()
        };

        await mongo.insertOne(acc);
        res.end(acc._id.toString());

        log.info('Created account \'%s\'', body.name);

        if(conf.hooks.onCreateAccount){
            delete acc.salt;
            delete acc.hash;
            post(conf.hooks.onCreateAccount, {
                'Content-Type': 'application/json',
                'X-Okaeri-Event': 'new-account'
            }, acc);
        }
    },

    async auth({ body, res, mongo }){
        let out = v.schema({ name: v.str, password: v.str }).test(body);
        res.badRequest(!out.ok, out.errs);

        let acc = await mongo.findOne({ name: body.name });
        res.unauthorized(!acc, 'Authentication failed');

        let claim = await hashPassword(body.password, acc.salt);
        res.unauthorized(claim !== acc.hash, 'Authentication failed');

        res.end(acc._id.toString());
    },

    async info({ headers, conf, params, res, mongo }){
        let id = params.acc || headers[conf.identityHeader];

        try{
            var acc = await mongo.findOne({ _id: new mongo.id(id) });
        }
        finally{
            res.notFound(!acc);
        }

        let out = { ...acc.metadata, id, name: acc.name, creation: acc.creation };
        res.set('Content-Type', 'application/json');
        res.end(JSON.stringify(out));
    }

}


lib/main.js
const mongo = require('nodecaf-mongo');
const crypto = require('crypto');
const v = require('valinor');

function generateSalt(){
    return new Promise( (resolve, reject) =>
        crypto.randomBytes(32, (err, buf) =>
            /* istanbul ignore next */
            err ? reject(err) : resolve(buf.toString('hex'))));
}

function hashPassword(password, salt){
    return new Promise( (resolve, reject) =>
        crypto.pbkdf2(password, salt, 1e5, 64, 'sha512', (err, buf) =>
            /* istanbul ignore next */
            err ? reject(err) : resolve(buf.toString('hex'))));
}

const ACC_PROJECTION = { hash: 0, salt: 0, _id: 0 };

module.exports = class Okaeri {

    constructor(conf){
        this.running = false;
        this.conf = {
            loginField: 'name',
            accountSchema: v.schema({
                name: v.str,
                password: v.str.between(8, 128)
            }),
            ...conf
        };
    }

    async start(){
        this.running = true;
        this.mongo = await mongo(this.conf.mongo, {
            account: 'Account',
            group: 'Group'
        })
        this.mongo.account.createIndexes({ [this.conf.loginField]: 1 }, { unique: true });
    }

    async stop(){
        this.running = false;
        await this.mongo.close();
    }

    async createAccount(input){

        let out = this.conf.accountSchema.test(input);
        if(!out.ok)
            return out;

        let lf = this.conf.loginField;
        if(await this.mongo.account.findOne({ [lf]: input[lf] }))
            return { ok: false, type: 'conflict' };

        let salt = await generateSalt();

        let acc = {
            ...out.final,
            salt,
            hash: await hashPassword(out.final.password, salt),
            groups: [],
            creation: new Date()
        };

        delete acc.password;

        await this.mongo.account.insertOne(acc);

        return { ok: true, id: acc._id.toString() };
    }

    async checkCredentials(key, password){
        let acc = await this.mongo.account.findOne(
            { [this.conf.loginField]: key },
            { hash: 1, salt: 1 }
        );

        if(!acc)
            return { ok: false, type: 'wrong' };

        let claim = await hashPassword(password, acc.salt);
        if(claim !== acc.hash)
            return { ok: false, type: 'wrong' };

        this.mongo.account.updateOne(
            { _id: acc._id },
            { $set: { lastAuth: new Date() } }
        );

        return { ok: true, id: acc._id.toString() };
    }

    async readAccount(id){
        var acc = await this.mongo.account.findOne({ _id: this.mongo.id(id) },
            { projection: ACC_PROJECTION });

        if(!acc)
            return { ok: false, type: 'unknown' };

        acc.id = id;

        return { ok: true, acc };
    }

    async updateAccount(id, data = {}){
        delete data.hash;
        delete data.salt;
        data.lastUpdate = new Date();

        // TODO logic if password changes
        // TODO logic if KEY changes
        delete data.password;

        let res = await this.mongo.account.updateOne({ _id: this.mongo.id(id) }, { $set: data });

        if(res.result.nModified == 0)
            return { ok: false, type: 'unknown' };

        return { ok: true };
    }

    async queryAccounts(filters = {}, order = false, page = 1){

        let result = await this.mongo.account.aggregate([
            { $sort: order || { name: 1 } },
            ...Object.keys(filters).length ? [{ $match: filters }] : [],
            { $skip: (page - 1) * 30 },
            { $limit: 30 },
            { $addFields: { id: '$_id'} },
            { $project: ACC_PROJECTION }
        ]);

        return { ok: true, accs: await result.toArray() };
    }

    async iterateAccounts(filters = {}){
        return await this.mongo.account.aggregate([
            ...Object.keys(filters).length ? [{ $match: filters }] : [],
            { $addFields: { id: '$_id'} },
            { $project: ACC_PROJECTION }
        ]);
    }

    async createGroup(input){
        let out = v.schema({ name: v.str, code: v.str }).test(input);
        if(!out.ok)
            return out;

        if(await this.mongo.group.findOne({ code: out.final.code }))
            return { ok: false, type: 'conflict' };

        let group = {
            ...out.final,
            accounts: [],
            creation: new Date()
        };

        await this.mongo.group.insertOne(group);

        return { ok: true, id: group._id.toString() };
    }

    async updateGroup(id, data = {}){
        let out = v.schema({ name: v.str.opt, code: v.str.opt }).test(data);
        if(!out.ok)
            return out;

        out.final.lastUpdate = new Date();

        let res = await this.mongo.group.updateOne({ _id: this.mongo.id(id) }, { $set: out.final });

        if(res.result.nModified == 0)
            return { ok: false, type: 'unknown' };

        return { ok: true };
    }

    async readGroup(id){
        var group = await this.mongo.group.findOne({ _id: this.mongo.id(id) });

        if(!group)
            return { ok: false, type: 'unknown' };

        let res = await this.mongo.account.aggregate([
            { $match: { _id: { $in: group.accounts } } },
            { $sort: { name: 1, _id: 1 } },
        ]);

        group.accounts = (await res.toArray()).map(a => {
            delete a.salt;
            delete a.hash;
            delete a.creation;
            delete a.groups;
            a.id = a._id.toString();
            delete a._id;
            return a;
        });

        return { ok: true, group };
    }

    async removeGroup(id){
        let _id = this.mongo.id(id);
        let g = await this.mongo.group.findOne({ _id });
        if(!g)
            return { ok: false, type: 'unknown' };

        await this.mongo.group.removeOne({ _id });

        return { ok: true, code: g.code }
    }

    async queryGroups(filters = {}, order = false, page = 1){

        let result = await this.mongo.group.aggregate([
            { $sort: order || { name: 1 } },
            ...Object.keys(filters).length ? [{ $match: filters }] : [],
            { $skip: (page - 1) * 30 },
            { $limit: 30 },
            { $addFields: { members: { $size: '$accounts' } } },
            { $project: { accounts: 0 } }
        ]);

        return { ok: true, groups: await result.toArray() };
    }

    async addAccountToGroup(accId, grpId){

        if(!await this.mongo.account.findOne({ _id: this.mongo.id(accId) }))
            return { ok: false, type: 'unknown' };

        if(!await this.mongo.group.findOne({ _id: this.mongo.id(grpId) }))
            return { ok: false, type: 'unknown' };

        await this.mongo.account.updateOne({ _id: this.mongo.id(accId) },
            { $addToSet: { groups: this.mongo.id(grpId) } });

        await this.mongo.group.updateOne({ _id: this.mongo.id(grpId) },
            { $addToSet: { accounts: this.mongo.id(accId) } });

        return { ok: true };
    }

    async removeAccountFromGroup(accId, grpId){

        if(!await this.mongo.account.findOne({ _id: this.mongo.id(accId) }))
            return { ok: false, type: 'unknown' };

        if(!await this.mongo.group.findOne({ _id: this.mongo.id(grpId) }))
            return { ok: false, type: 'unknown' };

        this.mongo.account.updateOne({ _id: this.mongo.id(accId) },
            { $pull: { groups: this.mongo.id(grpId) } });

        this.mongo.group.updateOne({ _id: this.mongo.id(grpId) },
            { $pull: { accounts: this.mongo.id(accId) } });

        return { ok: true };
    }

    async isAccountInAnyGroup(accId, groups){
        let result = await this.mongo.group.find({
            code: { $in: groups },
            accounts: this.mongo.id(accId)
        });

        return { ok: Boolean(await result.count()) };
    }

}
